src/Arcus.Messaging.Pumps.Abstractions/MessageHandling/MessageHandler.cs

```diff
@@ -26,11 +23,21 @@ private MessageHandler(Type serviceType, Func<object> createMessageHandlerImplem
         {
             Guard.NotNull(serviceType, nameof(serviceType));
             Guard.NotNull(createMessageHandlerImplementation, nameof(createMessageHandlerImplementation));
+            Guard.For<ArgumentException>(
+                () => serviceType.GenericTypeArguments.Length != 2, 
+                $"Message handler type '{serviceType.Name}' has not the expected 2 generic type arguments");
 
             _serviceType = serviceType;
             _createMessageHandlerImplementation = createMessageHandlerImplementation;
+
+            MessageType = _serviceType.GenericTypeArguments[0];
         }
 
+        /// <summary>
+        /// Gets the type of the message that this abstracted message handler can process.
+        /// </summary>
+        internal Type MessageType { get; }
```

Comment:
If it's internal, do you need it as a property?

Reply:
It's used in `MessagePump`. So we need access to it, but only internally.